source/fundamentals/monitoring.txt

```diff
@@ -0,0 +1,409 @@
+
+- :ref:`Monitor different types of events in the MongoDB Java Driver <monitoring-monitor-events>`
+- :ref:`Monitor connection pool events with Java Management Extensions (JMX) and JConsole <monitoring-jmx>`
+
+This guide shows how to use information about the activity of the driver in code.
+If you would like to learn how to record events the driver performs,
+consider reading our :doc:`guide on logging </fundamentals/logging>`.
+
+.. _monitoring-monitor-events:
+
+Monitor Events
+--------------
+
+To monitor an **event**, you must register a **listener** on your ``MongoClient``
+instance. 
+
+An event is any action that happens in a running program. The driver includes functionality
+for listening to a subset of the events that occur when the driver is running.
+
+A listener is a class that performs some action when certain events occur. All
+listener classes extend the ``EventListener`` interface. A listener's
+API defines the events it can respond to.
+
+Each method of a listener class represents a response to a certain event. Each
+method receives one argument: an object representing the event the method
+responds to. All event objects are a subclass of the ``EventObject`` class.  
+
+The MongoDB Java driver organizes the events it defines into three categories:
+
+- Command Events
+- Server Discovery and Monitoring Events
+- Connection Pool Events
+
+The following sections show how to monitor each event category.
+
+For a full list of the events you can monitor,
+:java-docs:`see the event package of the MongoDB Java Driver <apidocs/mongodb-driver-core/com/mongodb/event/package-summary.html>`.
+
+For more information on the classes and interfaces mentioned in this section,
+see the following API documentation:
+
+- `EventListener <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/EventListener.html>`__
+- `EventObject <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/EventObject.html>`__
+
+Command Events
+~~~~~~~~~~~~~~
+
+A command event is an event related to a MongoDB database command. Some
+examples of database commands that produce command events are ``find``,
+``insert``, ``delete``, and ``count``. 
+
+To monitor command events, write a class that implements the
+``CommandListener`` interface and register an instance of that class with your
+``MongoClient`` instance.
+
+For more information on MongoDB database commands, see the
+:manual:`MongoDB manual entry on database commands </reference/command/>`.
+
+.. note:: Redacted Output
+
+   As a security measure, the driver is unable to monitor some commands. This
+   protects the sensitive information contained in the output of these
+   commands. For a full list of commands the driver can't monitor,
+   see the 
+   :spec:`MongoDB command monitoring specification </command-monitoring/command-monitoring.rst#security>`.
+
+Example
+^^^^^^^
+
+This example shows how to make a timer for database commands. The timer makes it so
```

Comment:
Unfortunately this not a good example, because both CommandSucceededEvent and CommandFailedEvent include an elapsedTime property!

Reply:
Also, if this is all you want to do, better to enable DEBUG logging, which already logs elapsed time.

Reply:
[Facepalm](https://en.wikipedia.org/wiki/Facepalm)

Thanks for pointing this out. I agree that as there are much more straightforward ways to achieve this, this example should be changed. I have changed this example to be a command counter (keeps track of how many times each command has been successfully executed in a map and prints the map). Fix: https://github.com/mongodb/docs-java/pull/88/commits/c0fecb8671ddbe9318a92b1c759220198d4bfb04

Reply:
This is a better example.  Only suggestion is to make the `commandSucceeded` method` `synchronized`, since as is the listener will suffer from race conditions in a multi-threaded application.